internal/controller/workload_controller.go

```diff
@@ -90,11 +95,191 @@ func (aw *AppWrapper) PodSets() []kueue.PodSet {
+	podSetsInfoIndex := 0
+	for componentIndex := range aw.Spec.Components {
+		component := &aw.Spec.Components[componentIndex]
+		objChanged := false
+		obj := &unstructured.Unstructured{}
+		if _, _, err := unstructured.UnstructuredJSONScheme.Decode(component.Template.Raw, nil, obj); err != nil {
+			return err
+		}
+
+		for _, podSet := range component.PodSets {
+			podSetsInfoIndex += 1
+			if podSetsInfoIndex <= len(podSetsInfo) {
+				toInject := podSetsInfo[podSetsInfoIndex-1]
+				parts := strings.Split(podSet.Path, ".")
+				p := obj.UnstructuredContent()
+				var ok bool
+				for i := 1; i < len(parts); i++ {
+					p, ok = p[parts[i]].(map[string]interface{})
+					if !ok {
+						return fmt.Errorf("path element %v not found (segment %v of %v)", parts[i], i, len(parts))
+					}
+				}
+				objChanged = true // Even if currentInfo is empty, we will still add appWrapperLabel to p.metadata.labels
+
+				if _, ok := p["metadata"]; !ok {
+					p["metadata"] = make(map[string]interface{})
+				}
+				metadata := p["metadata"].(map[string]interface{})
+
+				// Annotations
+				if len(toInject.Annotations) > 0 {
+					if _, ok := metadata["annotations"]; !ok {
+						metadata["annotations"] = make(map[string]string)
+					}
+					annotations := metadata["annotations"].(map[string]string)
+					if err := utilmaps.HaveConflict(annotations, toInject.Annotations); err != nil {
+						return podset.BadPodSetsUpdateError("annotations", err)
+					}
+					metadata["annotations"] = utilmaps.MergeKeepFirst(annotations, toInject.Annotations)
+				}
+
+				// Labels
+				if _, ok := metadata["labels"]; !ok {
+					metadata["labels"] = make(map[string]string)
+				}
+				labels := metadata["labels"].(map[string]string)
+				labels[appWrapperLabel] = aw.Name
+				if len(toInject.Labels) > 0 {
+					if err := utilmaps.HaveConflict(labels, toInject.Labels); err != nil {
+						return podset.BadPodSetsUpdateError("labels", err)
+					}
+					labels = utilmaps.MergeKeepFirst(labels, toInject.Labels)
+				}
+				metadata["labels"] = labels
+
+				spec := p["spec"].(map[string]interface{}) // AppWrapper ValidatingAC ensures this succeeds
+
+				// NodeSelectors
+				if len(toInject.NodeSelector) > 0 {
+					if _, ok := spec["nodeSelector"]; !ok {
+						spec["nodeSelector"] = make(map[string]string)
+					}
+					selector := spec["nodeSelector"].(map[string]string)
+					if err := utilmaps.HaveConflict(selector, toInject.NodeSelector); err != nil {
+						return podset.BadPodSetsUpdateError("nodeSelector", err)
+					}
+					spec["nodeSelector"] = utilmaps.MergeKeepFirst(selector, toInject.NodeSelector)
+				}
+
+				// Tolerations
+				if len(toInject.Tolerations) > 0 {
+					if _, ok := spec["tolerations"]; !ok {
+						spec["tolerations"] = []interface{}{}
+					}
+					tolerations := spec["tolerations"].([]interface{})
+					for _, addition := range toInject.Tolerations {
+						bytes, err := json.Marshal(addition)
```

Comment:
since we know the type of `addition` it seems an explicit `map[string]interface{}` synthesis would be more efficient